unit_tests/test_grid_fee_caclulation.py

```diff
@@ -0,0 +1,151 @@
+                                }}
+               }}
+      }
+
+
+class TestGridFeeCalculation(unittest.TestCase):
+
+    def setUp(self):
+        self.grid_fee_calc = GridFeeCalculation()
+        self.grid_fee_calc.latest_grid_stats_tree = grid_stats_example
+        self.grid_fee_calc._populate_grid_fee_mappings()
+
+    @parameterized.expand([['last_market_fee'], ['next_market_fee']])
+    def test_grid_fee_is_calculated_correctly_for_leaf_devices(self, fee_type):
+        leaf_names = ['PV 2.1', 'PV 1.2', 'PV 3']
+        for leaf_name in leaf_names:
+            expected_fee = float(leaf_name.split(' ')[-1])
+            if fee_type == 'next_market_fee':
+                expected_fee += 1.
+            assert expected_fee == self.grid_fee_calc.calculate_grid_fee(leaf_name, fee_type=fee_type)
+
+    @parameterized.expand([['last_market_fee'], ['next_market_fee']])
+    def test_grid_fee_is_calculated_correctly_for_markets(self, fee_type):
+        market_names = ['House 2.1', 'House 1.2', 'Street 1', 'Grid 10']
+        for market_name in market_names:
+            expected_fee = float(market_name.split(' ')[-1])
+            if fee_type == 'next_market_fee':
+                expected_fee += 1.
+            assert expected_fee == self.grid_fee_calc.calculate_grid_fee(market_name, fee_type=fee_type)
+
+    @parameterized.expand([['last_market_fee'], ['next_market_fee']])
+    def test_grid_fee_is_calculated_correctly_for_leaf_to_grid(self, fee_type):
+        target_market = 'Grid 10'
+        if fee_type == 'next_market_fee':
+            leaf_name_expected_fee = {'PV 2.1': 17.1,
+                                      'PV 1.2': 15.2,
+                                      'PV 3': 15.}
+        else:
+            leaf_name_expected_fee = {'PV 2.1': 14.1,
+                                      'PV 1.2': 12.2,
+                                      'PV 3': 13.}
+        for leaf_name, expected_fee in leaf_name_expected_fee.items():
+            assert expected_fee == self.grid_fee_calc.calculate_grid_fee(
+                start_market_or_device_name=leaf_name,
+                target_market_or_device_name=target_market,
+                fee_type=fee_type)
+
+    @parameterized.expand([['last_market_fee'], ['next_market_fee']])
+    def test_grid_fee_is_calculated_correctly_for_leaf_to_leaf(self, fee_type):
+        if fee_type == 'next_market_fee':
+            leaf_name_expected_fee = {21.3: ['PV 2.1', 'Load 1.2'],
+                                      21.1: ['PV 2.1', 'Load 3'],
+                                      15.0: ['Market Maker', 'Load 3']}
+        else:
+            leaf_name_expected_fee = {16.3: ['PV 2.1', 'Load 1.2'],
+                                      17.1: ['PV 2.1', 'Load 3'],
+                                      13.0: ['Market Maker', 'Load 3']}
+        for expected_fee, leaf_names in leaf_name_expected_fee.items():
+            assert isclose(expected_fee, self.grid_fee_calc.calculate_grid_fee(
```

Comment:
I'm aware of how isClose works , but why would the values slightly differ at first place ?

Reply:
This happens some times when floating precision propagate in calculations. Here the calculated grid_fee was something like `12.999993` and the expected grid_fee was `13`